src/lender/tranche/v3/tranche.sol

```diff
@@ -0,0 +1,179 @@
+        return;
+    } 
+
+    uint delta = safeSub(currentRedeemAmount, newRedeemAmount);
+    if (delta > 0) {
+      require(token.transferFrom(self, msg.sender, delta), "token-transfer-failed");
+    }
+}
+
+// the disburse function can be used after an epoch is over to receive currency and tokens
+function disburse(uint epochID) public {
+    // require epoch is settled
+    require((epochs[epochID].tokenPrice > 0), "epoch-not-settled-yet");
+        
+    uint currencyAmount = calcCurrencyDisbursement(epochID);
+    uint tokenAmount = calcTokenDisbursement(epochID);
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = 0;
+    if (currencyAmount > 0) {
+        require(currency.transferFrom(self, msg.sender, currencyAmount), "currency-transfer-failed"); 
+    }
+    
+    epochs[epochID].redeemTokenAmount[msg.sender] = 0;
+    if (tokenAmount > 0) {
+        require(token.transferFrom(self, msg.sender, tokenAmount), "token-transfer-failed"); 
+    }
+}
+
+function calcCurrencyDisbursement(uint epochID) public view returns(uint) {
+    // currencyAmount = tokenAmount * percentage * tokenPrice 
+    uint currencyAmount = rmul(rmul(epochs[epochID].redeemTokenAmount[msg.sender], epochs[epochID].redeemFulfillment), epochs[epochID].tokenPrice);
+    // currencyAmount += unused dai from supply
+    return safeAdd(currencyAmount, rmul(epochs[epochID].supplyCurrencyAmount[msg.sender], safeSub(ONE, epochs[epochID].supplyFulfillment)));
+}
+
+function calcTokenDisbursement(uint epochID) public view returns(uint) {
+    // take currencyAmount from redeemOrder
+    uint tokenAmount = rdiv(rmul(epochs[epochID].supplyCurrencyAmount[msg.sender], epochs[epochID].supplyFulfillment), epochs[epochID].tokenPrice);
+    // add leftovers from supplies
+    return safeAdd(tokenAmount, rmul(epochs[epochID].redeemTokenAmount[msg.sender], safeSub(ONE, epochs[epochID].redeemFulfillment)));
+}
+
+// called by epoch coordinator in epoch execute method
+function epochUpdate(uint epochID, uint supplyFulfillment_, uint redeemFulfillment_, uint tokenPrice_) public auth {
+    epochs[epochID].supplyFulfillment = supplyFulfillment_;
+    epochs[epochID].redeemFulfillment = redeemFulfillment_;
+    epochs[epochID].tokenPrice = tokenPrice_;
+     
+    // burn amount of tokens for that epoch
+    uint burnAmount = rmul(epochs[epochID].totalRedeem, epochs[epochID].redeemFulfillment);
+
+    // mint amount of tokens for that epoch
+    uint mintAmount = rdiv(rmul(epochs[epochID].totalSupply, epochs[epochID].supplyFulfillment), epochs[epochID].tokenPrice);
+    // burn tokens that are not needed for disbursement
+    if (burnAmount > mintAmount) {
+        uint diff = safeSub(burnAmount, mintAmount);
+        token.burn(self, diff);
+        return;
+    }
+
+    // mint tokens that are required for disbursement
+    uint diff = safeSub(mintAmount, burnAmount);
+    if (diff > 0) {
+        token.mint(self, diff);
+    }
+}
```

Comment:
What is currently still missing is the currency take from reserve or give to reserve depending on the currency delta

Reply:
true!